### ISSUE TITLE
ledger-tool verify is hung up for a while waiting for AccountBackgroundService

### ISSUE THREAD

**Issue by steviez** [anza-xyz/agave#346]
#### Problem
Running `agave-ledger-tool verify` with a recent commit near the tip of master sees a non-trivial slowdown after tx-processing has finished. Namely, it appears to be waiting on ABS:
```
[2024-03-20T20:04:09.010886812Z INFO  solana_ledger::blockstore_processor] ledger processed in 2 ms, ...
...
// Exit flag set which should propagate to AHV and ABS
[2024-03-20T20:04:09.407809319Z INFO  solana_core::accounts_hash_verifier] AccountsHashVerifier has stopped
...
[2024-03-20T20:11:42.695623647Z INFO  solana_runtime::accounts_background_service] AccountsBackgroundService has stopped
```
Note the timestamp difference; AHV returns almost immediately whereas ABS stops `> 7` minutes later. Here are all the logs inbetween:
```
[2024-03-20T20:04:08.865586216Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash_scan=1i
[2024-03-20T20:04:09.407809319Z INFO  solana_core::accounts_hash_verifier] AccountsHashVerifier has stopped
[2024-03-20T20:05:03.132225735Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash_scan=0i
[2024-03-20T20:05:03.156931563Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash_dedup=1i
[2024-03-20T20:05:38.739498389Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash_dedup=0i
[2024-03-20T20:05:38.752142142Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash_merkle_tree=1i
[2024-03-20T20:05:40.891854054Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash_merkle_tree=0i
[2024-03-20T20:05:44.359742882Z INFO  solana_accounts_db::accounts_db] calculate_accounts_hash_from_storages: slot: 254911099, Full(AccountsHash(EiKKH9o7NGgi7jw9MSAxidLtidhUx61ZLMWk88KeoMUT)), capitalization: 572149832212208582
[2024-03-20T20:05:45.781712115Z INFO  solana_metrics::metrics] datapoint: cache_hash_data_stats cache_file_size=68760899328i cache_file_count=175i total_entries=955012452i loaded_from_cache=174i saved_to_cache=174i entries_loaded_from_cache=477506226i save_us=46980731i write_to_mmap_us=46620189i create_save_us=312311i load_us=0i read_us=3231i unused_cache_files=175i hits=0i misses=174i
[2024-03-20T20:05:45.782159006Z INFO  solana_metrics::metrics] datapoint: calculate_accounts_hash_from_storages total_us=96919324i mark_time_us=0i cache_hash_data_us=1510i accounts_scan_us=54266994i eliminate_zeros_us=35582547i hash_us=2139694i sort_us=19543420i hash_total=443043076i storage_sort_us=29499i collect_snapshots_us=73647i num_snapshot_storage=407166i scan_chunks=174i num_slots=407166i num_dirty_slots=0i storage_size_min=311296i storage_size_quartile_1=417792i storage_size_quartile_2=462848i storage_size_quartile_3=569344i storage_size_max=343113728i storage_size_avg=573478i roots_older_than_epoch=124i oldest_root=254478968i longest_ancient_scan_us=0i sum_ancient_scans_us=0i count_ancient_scans=0i append_vec_sizes_older_than_epoch=63826288i accounts_in_roots_older_than_epoch=142832i pubkey_bin_search_us=56003607i
[2024-03-20T20:05:45.782165879Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash=0i
[2024-03-20T20:05:45.786250783Z INFO  solana_runtime::bank] Initial background accounts hash verification has stopped
[2024-03-20T20:11:41.313515479Z INFO  solana_accounts_db::accounts_db] remove_dead_slots_metadata: 122 dead slots
[2024-03-20T20:11:41.315763717Z INFO  solana_metrics::metrics] datapoint: clean_purge_slots_stats safety_checks_elapsed=1i remove_cache_elapsed=0i remove_storage_entries_elapsed=431i drop_storage_entries_elapsed=1i num_cached_slots_removed=0i num_stored_slots_removed=122i total_removed_storage_entries=122i total_removed_cached_bytes=0i total_removed_stored_bytes=62839152i scan_storages_elapsed=0i purge_accounts_index_elapsed=0i handle_reclaims_elapsed=0i
[2024-03-20T20:11:41.315805737Z INFO  solana_metrics::metrics] datapoint: accounts_index_roots_len roots_len=407044i uncleaned_roots_len=1i roots_range_width=432018i unrooted_cleaned_count=0i rooted_cleaned_count=122i clean_unref_from_storage_us=80350i clean_dead_slot_us=56i append_vecs_open=407166i
[2024-03-20T20:11:41.315811128Z INFO  solana_metrics::metrics] datapoint: clean_accounts total_us=452307833i collect_delta_keys_us=43913i oldest_dirty_slot=254911099i pubkeys_removed_from_accounts_index=2331i dirty_ancient_stores=0i dirty_store_processing_us=49750888i accounts_scan=22879414i clean_old_rooted=342002459i store_counts=10647970i purge_filter=2555782i calc_deps=7211465i reclaims=182801i delta_insert_us=4282495i delta_key_count=17541843i dirty_pubkeys_count=0i sort_us=9363089i useful_keys=17541843i total_keys_count=17541843i scan_found_not_zero=9058004i scan_not_found_on_fork=0i scan_missing=0i uncleaned_roots_len=407073i clean_old_root_us=1943283i clean_old_root_reclaim_us=339994628i reset_uncleaned_roots_us=25242i remove_dead_accounts_remove_us=339772510i remove_dead_accounts_shrink_us=223505i clean_stored_dead_slots_us=101696i roots_added=407166i roots_removed=122i active_scans=0i max_distance_to_min_scan_slot=0i ancient_account_cleans=0i next_store_id=407166i
[2024-03-20T20:11:41.379356012Z INFO  solana_metrics::metrics] datapoint: accounts_db_active clean=0i
[2024-03-20T20:11:42.595493256Z INFO  solana_metrics::metrics] datapoint: accounts_db_active shrink=1i
[2024-03-20T20:11:42.595545706Z INFO  solana_metrics::metrics] datapoint: accounts_db_active shrink=0i
[2024-03-20T20:11:42.595553822Z INFO  solana_metrics::metrics] datapoint: accounts_background_service duration_since_previous_submit_us=453587771i num_iterations=1i cumulative_runtime_us=453587771i mean_runtime_us=453587771i min_runtime_us=453587771i max_runtime_us=453587771i
[2024-03-20T20:11:42.695623647Z INFO  solana_runtime::accounts_background_service] AccountsBackgroundService has stopped
```

A few items jump out:
- `accounts_db_active hash_scan=0i` comes in after about 60 seconds, so this seemingly doesn't bail out on exit flag
- `Initial background accounts hash verification has stopped` comes in after about 90 seconds
- There is a large gap after above item, but before next log line `remove_dead_slots_metadata` ... like 6 minutes

#### Proposed Solution
Thought 1:
My possibly naive understanding is that `ledger-tool` should be able to opt out of these items; maybe we want some of these steps to occur for case of `create-snapshot`. Or, maybe opt-in if there is debug value.

Thought 2:
A step further, I think it would be preferred to plumb the `exit` flag down in these routines so that if someone does set the exit flag, they are aborted in a timely manner. Doing this may remove the need for any special handling to skip stuff in the ledger-tool case. If we go this route, we probably need a sanity check on whether there are any instances in which we should let these operations complete for ledger-tool. Again, `create-snapshot` command comes to mind as a candidate

**Comment by steviez** [comment]
@brooksprumo and @jeffwashington - Any ideas from you two on this ?

**Comment by jeffwashington** [comment]
accounts_db_active clean=0i

This means clean was running all this time.
total_us=452,307,833

we do not have a exit flag running all the way down into clean.

the hash calc ended not too long.

**Comment by steviez** [comment]
I see this:
```
[2024-03-20T20:04:08.751288211Z INFO  solana_metrics::metrics] datapoint: verify_snapshot_bank clean_us=2i shrink_us=1i verify_accounts_us=112i verify_bank_us=1485i
```
So I think we are skipping clean for the snapshot bank and something else must be triggering another clean

**Comment by jeffwashington** [comment]
> something else must be triggering another clean

accounts background service is triggering a clean. The initial clean is skipped. the initial clean is synchronous and has to finish before replay starts. That is almost always skipped.

**Comment by steviez** [comment]
Is propagating an exit flag and potentially bailing on a partial clean safe ? I can think of two cases:
- `validator` is running and someone runs `agave-validator exit`. If we quit on a clean midway through, is there any state that is now in an in-between / bad state when process is restarted (ie fastboot)?
    - FWIW, if all of the threads don't join in 5 seconds, we bail out with `std::process:exit(0)` anyways so this is something we'd potentially want to address regardless.
- `ledger-tool` finishes processing and returns to "command" logic. In some instances, we are done. But in others, such as the `accounts` subcommand, interacting with AccountsDb further may be desired

If respecting the `exit` flag is safe for both cases, that seems like the easiest solution. It looks like `AccountsDb::new_with_config()` already takes the `Arc<AtomicBool>` exit flag, so we'd just need to clone it and retain a copy in `AccountsDb`

**Comment by steviez** [comment]
@jeffwashington @brooksprumo - Any thoughts on my last question ? This is still a non-trivial slowdown, so it plumbing the `exit` flag down is safe, I'm willing to make the change to shave time off ledger-tool runs.

**Comment by jeffwashington** [comment]
I'm fine to plumb exit flag through.

**Comment by brooksprumo** [comment]
I think we could also skip `clean` and `shrink` for `ledger-tool verify`. Especially if there are not a lot of blocks to replay.

**Comment by jeffwashington** [comment]
we could pass osmething to ledger tool to have it nuke the initial clean/shrink data that it computed. This would cause the initial clean/shrink to effectlively be skipped.

**Comment by jeffwashington** [comment]
https://github.com/anza-xyz/agave/blob/9cd90751f0105665328691b2cf72d2a8310b09ed/accounts-db/src/accounts_db.rs#L9013-L9014

That may be enough. If we don't store that, it may think there is nothing to do on clean. @steviez can you check?

**Comment by steviez** [comment]
> https://github.com/anza-xyz/agave/blob/9cd90751f0105665328691b2cf72d2a8310b09ed/accounts-db/src/accounts_db.rs#L9013-L9014
> 
> That may be enough. If we don't store that, it may think there is nothing to do on clean. @steviez can you check?

To confirm, you just want me to comment out this line as a quick check, and if it avoids the clean, then we'll plumb a change more properly ?

**Comment by jeffwashington** [comment]
> > https://github.com/anza-xyz/agave/blob/9cd90751f0105665328691b2cf72d2a8310b09ed/accounts-db/src/accounts_db.rs#L9013-L9014
> > 
> > That may be enough. If we don't store that, it may think there is nothing to do on clean. @steviez can you check?
> 
> To confirm, you just want me to comment out this line as a quick check, and if it avoids the clean, then we'll plumb a change more properly ?

yes. hack it out. see if it solves the problem for your use case.
There may be one more place (uncleaned pubkeys somehow?).
If this solves the problem, then a ledger-tool cli arg could push a config arg down to this fn or somehwere near here where we effectively enable `for_ledger_tool_or_testing_avoid_initial_clean_or_whatever_we_call_it`. For a real validator, we don't need to be able to stop every loop in clean, shrink, ancient pack, etc. Doing so adds overhead, noise, and risk. If the whole goal is for ledger-tool to end more quickly, then avoiding seeding the initial clean may be the right approach.

**Comment by steviez** [comment]
> For a real validator, we don't need to be able to stop every loop in clean, shrink, ancient pack, etc. Doing so adds overhead, noise, and risk.

I'm not as sure about this. I think the overhead aspect is valid and should be considered, but I also think having the validator being able to stop in a timely manner when `validator exit` command is set. Currently, we just do a hard `std::process::exit()` ... is the hard exit potentially problematic / could anything get into corrupted state because of this?

In general, I think there is benefit in having all of our long running operations be abortable

**Comment by brooksprumo** [comment]
> Currently, we just do a hard `std::process::exit()` [..]

When we do `solana-validator exit`, we set the `exit` Atomic flag to true. But, if we don't stop fast enough, after 5 seconds we do `std::process::exit()`.

https://github.com/anza-xyz/agave/blob/e3e0ac1a3b9473f383759331f9f57c6a369e9311/validator/src/admin_rpc_service.rs#L252-L275

I recently used `solana-validator exit`, and I think it shutdown gracefully? Here's the logs. Nothing was snipped in between the "validator exit requested" and the "Starting validator" lines. Is my interpretation of the logs correct? I would've expected a "validator exit timeout" log otherwise.

```
[2024-03-27T16:47:42.402677525Z WARN  agave_validator::admin_rpc_service] validator exit requested
[2024-03-27T16:47:42.403917794Z INFO  solana_runtime::accounts_background_service] AccountsBackgroundService has stopped
[2024-03-27T16:47:42.563409198Z INFO  solana_ledger::blockstore_cleanup_service] BlockstoreCleanupService has stopped
[2024-03-27T16:48:03.821747719Z INFO  agave_validator] agave-validator 2.0.0 (src:e3e0ac1a; feat:1030785117, client:Agave)
[2024-03-27T16:48:03.821784710Z INFO  agave_validator] Starting validator with: ArgsOs {
[..]
```

**Comment by jeffwashington** [comment]
I would separate these concerns.

1. ledger tool doesn't stop. this impact developers and efficiency. This has an easy fix with no risk. This is the point of this issue.
2. we'd like long running bg processes to stop in less than 5 seconds in a running validator.

#1 seems easy to solve.
#2 would need to be scheduled and traded off risk vs. reward. and opportunity cost. To me, this is not worth it to spend time on. It is a nice to have. System state is not affected if these threads (clean, etc.) are killed in the middle of an operation.

**Comment by steviez** [comment]
> Nothing was snipped in between the "validator exit requested" and the "Starting validator" lines. Is my interpretation of the logs correct? I would've expected a "validator exit timeout" log otherwise.

Yes, I agree with your interpretation. But, note that this could be an intermittent issue; ie, whether you are in the middle of a long-running operation when you exit.

> I would separate these concerns.

Sure, I agree that they are separate but if they had a common solution then why not get both.

> System state is not affected if these threads (clean, etc.) are killed in the middle of an operation.

Great, I asked this question a bunch of message (may not have been clear that I was asking) up but didn't get a direct answer (or I might have missed it). In any case, if a killing these inflight operations is safe, then I agree with the points you made in your second bullet and that we can leave things as-is for now

When I get the chance, I will give @jeffwashington's suggestion in https://github.com/anza-xyz/agave/issues/346#issuecomment-2023127208 a shot and report back. Thanks for the insights Jeff and Brooks !

**Comment by alessandrod** [comment]
> @brooksprumo 
> I think we could also skip clean and shrink for ledger-tool verify. Especially if there are not a lot of blocks to replay.

When profiling the program runtime it's useful if other things (like ABS) do their thing as usual, so one can see the impact on replay threads.

> @steviez 
> But, note that this could be an intermittent issue; ie, whether you are in the middle of a long-running operation when you exit.

This is what's happening. With a mnb ledger from last week, ~10k slots, the initial clean takes ~4m in which ABS is inside clean_accounts() and doesn't get a chance to look at the exit flag. Slot replay takes ~30s, so the remaining time we wait for clean_accounts() to return. Plumbing the exit flag down a few more layers where clean_account() is looping would fix it.

Alternatively ledger-tool could _not_ join the ABS threads, and let it die when blockstore processing ends (this is what I'm doing locally currently).

**Comment by jeffwashington** [comment]
we could add a ledger tool flag to not populate the initial clean with duplicates. This would cause clean and shrink to run but not find anything to do.

**Comment by alessandrod** [comment]
> we could add a ledger tool flag to not populate the initial clean with duplicates. This would cause clean and shrink to run but not find anything to do.

Yeah this sounds like the simplest and less risky solution

**Comment by steviez** [comment]
Will be testing this later today BTW

**Comment by steviez** [comment]
Commenting out these lines did not avoid the ABS slowdown:
https://github.com/anza-xyz/agave/blob/9cd90751f0105665328691b2cf72d2a8310b09ed/accounts-db/src/accounts_db.rs#L9013-L9014

> When profiling the program runtime it's useful if other things (like ABS) do their thing as usual, so one can see the impact on replay threads.

In light of Alessandro's comment here, it seems slightly more difficult to pick the right strategy:
- If we're only unpacking the snapshot and not replaying any blocks, we currently attempt to exit immediately and wait on ABS for a while
- If we're replaying 100's or 1000's of slots, the clean will occur while we're replaying. This probably slows down replay for `ledger-tool`, but makes things more comparable to `validator` and also means we likely won't have any extra wait once tx processing has finished

So yeah, I agree that a flag seems like the right move. @jeffwashington - Have anything else I should try? I can poke around a bit if not, but you are also much more familiar with this code than me

**Comment by jeffwashington** [comment]
> Commenting out these lines did not avoid the ABS slowdown:

@steviez can you send me the logs after ledger processing stopped? In theory the things that ran then would either respect the exit flag or would not take long.

**Comment by steviez** [comment]
I posted in the PR, but maybe should have posted here first. In any case, the linked PR that checks the exit flag more frequently did not help. The issue is a clean starts and is taking a while:
```
// Snapshot bank first frozen
[2024-05-29T17:15:02.354026793Z INFO  solana_runtime::bank] bank frozen: 268499996 ...

[2024-05-29T17:15:02.444948540Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash=1i
[2024-05-29T17:15:02.446397061Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash_scan=1i
[2024-05-29T17:15:02.872101391Z INFO  solana_metrics::metrics] datapoint: accounts_db_active flush=1i
[2024-05-29T17:15:02.872113364Z INFO  solana_metrics::metrics] datapoint: accounts_db_active flush=0i
[2024-05-29T17:15:02.872115558Z INFO  solana_metrics::metrics] datapoint: accounts_db_active clean=1i

[2024-05-29T17:17:23.121183654Z INFO  solana_ledger::blockstore_processor] ledger processed in 2 minutes, 20 seconds, 246 ms, ...

[2024-05-29T17:17:42.029377762Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash_scan=0i
[2024-05-29T17:17:42.066207558Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash_dedup=1i
[2024-05-29T17:18:00.965311654Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash_dedup=0i
[2024-05-29T17:18:00.980008602Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash_merkle_tree=1i
[2024-05-29T17:18:02.218567428Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash_merkle_tree=0i
[2024-05-29T17:18:13.628383972Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash=0i
[2024-05-29T17:18:13.661989984Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash=1i
[2024-05-29T17:18:13.663247821Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash_scan=1i
[2024-05-29T17:18:19.393629231Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash_scan=0i
[2024-05-29T17:18:19.393751724Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash_dedup=1i
[2024-05-29T17:18:20.100557640Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash_dedup=0i
[2024-05-29T17:18:20.105439944Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash_merkle_tree=1i
[2024-05-29T17:18:20.146674334Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash_merkle_tree=0i
[2024-05-29T17:18:21.851065346Z INFO  solana_metrics::metrics] datapoint: accounts_db_active hash=0i
[2024-05-29T17:25:31.015283032Z INFO  solana_metrics::metrics] datapoint: accounts_db_active clean=0i
```

**Comment by steviez** [comment]
Another idea - maybe we do something similar to `validator` in that we send the exit flag but do NOT block on joining ABS thread here:
https://github.com/anza-xyz/agave/blob/2de7b565e8b1101824a5e3bac74f3a8cce88ea72/ledger-tool/src/ledger_utils.rs#L407

Instead, we pass ABS thread handle to a new thread, and join ABS within that background thread. Then, we go about our business doing whatever else is desired in ledger-tool and if we get to the end and things haven't all returned, we just bail out. Again, validator does something similar as seen below when `validator exit` is used:

https://github.com/anza-xyz/agave/blob/2de7b565e8b1101824a5e3bac74f3a8cce88ea72/validator/src/admin_rpc_service.rs#L252-L275

Our exact solution would obviously be slightly; we could probably just use [`JoinHandle::is_finished()`](https://doc.rust-lang.org/std/thread/struct.JoinHandle.html#method.is_finished) and log warning + exit if this "background thread" has not yet finished

It'd be good to here if the AccountsDb folks had any concerns (race condition with create-snapshot ?) with this approach

**Comment by alessandrod** [comment]
> Instead, we pass ABS thread handle to a new thread, and join ABS within that background thread

I think we can just not join at all?

**Comment by brooksprumo** [comment]
With `ledger-tool create-snapshot`, it calls `snapshot_bank_utils::bank_to_full_snapshot_archive()` to create the snapshot. This doesn't use the background services (ABS, AHV, SPS). So once `bank_to_snapshot_archive()` completes, it should be fine to just exit out and not join ABS/AHV/SPS.
